AtlasEngine: Scale glyphs to better fit the cell size

lhecker opened the pull request:
This commit contains 3 improvements for glyph rendering:
* Scale block element and box drawing characters to fit the cell size
  "perfectly" without leaving pixel gaps between cells.
* Use `IDWriteTextLayout::GetOverhangMetrics` to determine whether glyphs
  are outside the given layout box and if they are, offset their position
  to fit them back in. If that still fails to fit, we downscale them.
* Always scale up glyphs that are more than 2 cells wide
  This ensures that long ligatures that mimic box drawing characters like
  "===" under Cascadia Code are upscaled just like regular box drawings.
  Unfortunately this results in ligature-heavy text (like Myanmar) to get an
  "uneven" appearance because some ligatures can suddenly appear too large.
  It's difficult to come up with a good heuristic here.

Closes #12512

## Validation Steps Performed
* Print UTF-8-demo.txt
* Block characters don't leave gaps ✅
* Print a lorem-ipsum in Myanmar
* Glyphs aren't cut off anymore ✅
* Print a long "===" ligature under Cascadia Code
* The ligature is as wide as the number of cells used ✅

lhecker commented on src/renderer/atlas/AtlasEngine.cpp:
```diff
@@ -1046,9 +1047,11 @@ void AtlasEngine::_recreateFontDependentResources()
+                if (const auto textFormat3 = textFormat.try_query<IDWriteTextFormat3>())
                 {
-                    if (const auto textFormat3 = textFormat.try_query<IDWriteTextFormat3>())
+                    textFormat3->SetAutomaticFontAxes(DWRITE_AUTOMATIC_FONT_AXES_OPTICAL_SIZE);
```
This change might be unrelated, but I felt like that this still fits in the theme of the PR.
It ensures that a good "optical size" is chosen for variable width fonts which I feel like might impact overhangs.

lhecker commented on src/renderer/atlas/AtlasEngine.r.cpp:
```diff
@@ -266,23 +267,177 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
+    // Inversely we need to scale them up (not down) if they don't fill the box.
+    if (charsLength == 1 && key->chars[0] >= 0x2500 && key->chars[0] <= 0x259F)
+    {
+        // TODO: This code doesn't quite work yet unfortunately.
```
The TODO-issue is #12512.

lhecker commented:
Does this look correct to you? Because I can't replicate the exact same look you get given that config. For instance your second row of text is thin, but you specified that you set the font to extra-bold. 🤔
```json
{
    "antialiasingMode": "aliased",
    "experimental.useAtlasEngine": true,
    "font": 
    {
        "face": "Cascadia Code",
        "size": 17,
        "weight": "extra-bold"
    }
}
```

But I believe the fix is to:
* Add 0.5px of "margin" on each side when downscaling glyphs. That way we ensure that when DirectWrite/D2D snap glyph outlines to pixels, that those pixels actually end up within the layout box.
* `AtlasEngine::_resolveFontMetrics` seems to not respect font weight. When I change Cascadia Code from Regular to Bold the cell size doesn't change at all. But it should - the bold "M" character is a few pixels fatter after all.

lhecker commented:
> 1. font size does not match between two modes. Sorry, this may not be related, but maybe it is. In the following table I listed sizes which visually match with each other in the respective modes.

Yeah I did this intentionally. The current text renderer (DxEngine) draws glyphs with fractional font sizes so that they fit perfectly into their glyph cells. This solves some issues (glyphs like "========" in Cascadia Code not having the correct size in AtlasEngine) but brings other issues (glyphs being blurry as you noted). It's a trade-off and there's unfortunately no correct decision. 🙁 I went with the "pixel-perfect" approach which makes ASCII and CJK text look nicer.

The latest commit should make your arrows look a bit nicer. Unfortunately they're sometimes still slightly cut off but at least for me it's tolerable now... We'll have to see how it goes! 🙂

lhecker commented on src/renderer/atlas/AtlasEngine.r.cpp:
```diff
@@ -266,23 +275,207 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
+        if (cellCount > 2)
+        {
+            scalingRequired = true;
+            scale = _r.fontMetrics.scale;
```
I just realized that this approach is slightly incorrect.
TODO (for myself): Set `scale` to the `max(.x, .y)` so that the glyph is resized while preserving the aspect ratio.

zadjii-msft commented on src/renderer/atlas/DWriteTextAnalysis.cpp:
```diff
@@ -0,0 +1,172 @@
+#include "pch.h"
```
note to reviewers: this whole file is just code that was deleted from the `TextAnalyzer` in `AtlasEngine.cpp`

zadjii-msft commented on src/renderer/atlas/AtlasEngine.r.cpp:
```diff
@@ -271,23 +271,219 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
         _r.d2dRenderTarget->Clear();
+        if (scalingRequired)
+        {
+            // TODO: dx/dy should technically be pixel snapped
```
Is this TODO now or TODO later?

lhecker replied:
I wrote that comment for future-someone and not specifically for us.
I'll try to figure out the correct approach for this before it gets merged, but if I don't it's not that important anyways - the scaled glyphs are pretty crisp either way.

zadjii-msft commented on src/renderer/atlas/AtlasEngine.r.cpp:
```diff
@@ -245,19 +244,20 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
     }
 
-    auto options = D2D1_DRAW_TEXT_OPTIONS_CLIP;
+    auto options = D2D1_DRAW_TEXT_OPTIONS_NONE;
```
no more `_CLIP`?

lhecker replied:
Yeah it doesn't work with scaling.

zadjii-msft reviewed (commented):
All in all, the code looks good to me, just a couple questions before I give the ✅

DHowett commented on src/renderer/atlas/AtlasEngine.cpp:
```diff
@@ -966,9 +838,12 @@ void AtlasEngine::_recreateFontDependentResources()
 
         _r.cellSizeDIP.x = static_cast<float>(_api.fontMetrics.cellSize.x) / scaling;
         _r.cellSizeDIP.y = static_cast<float>(_api.fontMetrics.cellSize.y) / scaling;
+        _r.fontMetrics = _api.fontMetrics;
```
duplicate line with R844

DHowett commented on src/renderer/atlas/AtlasEngine.h:
```diff
@@ -992,7 +998,7 @@ namespace Microsoft::Console::Render
             std::vector<DWRITE_FONT_FEATURE> fontFeatures; // changes are flagged as ApiInvalidations::Font|Size
             std::vector<DWRITE_FONT_AXIS_VALUE> fontAxisValues; // changes are flagged as ApiInvalidations::Font|Size
-            FontMetrics fontMetrics; // changes are flagged as ApiInvalidations::Font|Size
+            FontMetrics fontMetrics; // changes are flagged as ApiInvalidations::Font
```
not Size?

DHowett reviewed (approved):
I'm cool with this, let's get it into selfhosting!

DHowett commented on src/renderer/atlas/AtlasEngine.r.cpp:
```diff
@@ -408,28 +405,34 @@ void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
+        // But we do want it technically, so this re-implements baseline snapping... I think?
+        // It calculates the new `baseline` height after transformation by `scale.y` relative to the center point `halfSize.y`.
+        //
+        // This works even if `scale.y == 1`, because then `baseline == baselineInDIP + offset.y` and `baselineInDIP`
```
to some extent i just have to, like... trust your math lol

lhecker replied:
Yeah, this is really hard to read:
```
baseline = halfSize.y + (_r.fontMetrics.baselineInDIP + offset.y - halfSize.y) * scale.y
```

If it helps here's an image explanation:
![image](https://user-images.githubusercontent.com/2256941/181860430-5beed2f0-df00-44cb-8f44-13c7111bec10.png)

`offset.y` shifts the glyph (and thus the baseline) up and down as needed to fit the glyph into the cell. So it's a simple `+ offset.y` and you arrive at my complex line of code there.

lhecker commented:
@zadjii-msft FYI I just pushed some commits. It addresses these issues:
* I (hopefully correctly) fixed support for baseline snapping - CJK looks a better now!
* Fixed centering of glyphs - the `offset` was applied to the `origin` _before_ the transformation was applied, which meant that the `offset` was applied twice: Once to center the origin of `DrawTextLayout` and once to offset the scaling center (= after scaling the glyph isn't centered anymore).
* The use `targetSize` was entirely incorrect - glyphs like "W" are just as wide as "M" (= cell width), but since `targetSize` is 1px smaller than the cell size, it accidentally downsized all W by a factor of like 0.999995.
